### ISSUE TITLE
selecting the time step for gradient truncation

### ISSUE THREAD

**Issue by daewon88** [mihirp1998/AlignProp#15]
Hi! Thank you for sharing your valuable work.

In the code, selecting the timestep for gradient truncation occurs every denoising step (lines 483-485). However, if the intention is to sample the truncation timestep from U(0,50), there may be some issues with this approach. Therefore, I suggest that selecting the timestep for gradient truncation should happen before each sampling, rather than each denoising step. Do you have any particular reason for this choice?

Thank you :)

```
if config.truncated_backprop:
    if config.truncated_backprop_rand:
        timestep = random.randint(config.truncated_backprop_minmax[0],config.truncated_backprop_minmax[1])
        if i < timestep:
            noise_pred_uncond = noise_pred_uncond.detach()
            noise_pred_cond = noise_pred_cond.detach()
```

**Comment by ajaysub110** [comment]
I had the same question while going through the code. @mihirp1998 do you have a particular reason for choosing this approach over what @daewon88 suggests? Thanks!

**Comment by mihirp1998** [comment]
Yes sorry, i didn't realise i was doing this. This would be conceptually wrong, although i don't expect the results to change, as it will be similar to sampling the truncation timestep outside the loop.

**Comment by mihirp1998** [comment]
Thanks for catching this, this is not representative of the report we have. This is not U(0,50), this results in giving more weight to the timesteps that are more closer to 50, it's more like an exponential decay in probablity as we move from 50 to 0.  Although i think emperically this would work better than U(0,50) as it regularizes the full backprop.

**Comment by mihirp1998** [comment]
Thanks for catching this bug! 

The current code is not using U(0,50), but it is instead a gaussian distribution that is centered at 42, i haven't ablated this with U(0,50), but once i do i'll  add this as an option in the code. 

![Figure_1](https://github.com/user-attachments/assets/11c1f20e-8699-458f-82bf-f8210302709e)
